auth/caching_sha2.go

```diff
@@ -0,0 +1,208 @@
+
+	return ret
+}
+
+func sha256crypt(plaintext string, salt []byte, iterations int) string {
+	// 1, 2, 3
+	tmpA := sha256.New()
+	tmpA.Write([]byte(plaintext))
+	tmpA.Write(salt)
+
+	// 4, 5, 6, 7, 8
+	tmpB := sha256.New()
+	tmpB.Write([]byte(plaintext))
+	tmpB.Write(salt)
+	tmpB.Write([]byte(plaintext))
+	sumB := tmpB.Sum(nil)
+
+	// 9, 10
+	var i int
+	for i = len(plaintext); i > MIXCHARS; i -= MIXCHARS {
+		tmpA.Write(sumB[:MIXCHARS])
+	}
+	tmpA.Write(sumB[:i])
+
+	// 11
+	for i = len(plaintext); i > 0; i >>= 1 {
+		if i%2 == 0 {
+			tmpA.Write([]byte(plaintext))
+		} else {
+			tmpA.Write(sumB)
+		}
+	}
+
+	// 12
+	sumA := tmpA.Sum(nil)
+
+	// 13, 14, 15
+	tmpDP := sha256.New()
+	for range []byte(plaintext) {
+		tmpDP.Write([]byte(plaintext))
+	}
+	sumDP := tmpDP.Sum(nil)
+
+	// 16
+	p := []byte{}
+	for i = len(plaintext); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			p = append(p, sumDP...)
+		} else {
+			p = append(p, sumDP[0:i]...)
+		}
+	}
+
+	// 17, 18, 19
+	tmpDS := sha256.New()
+	for i = 0; i < 16+int(sumA[0]); i++ {
+		tmpDS.Write(salt)
+	}
+	sumDS := tmpDS.Sum(nil)
+
+	// 20
+	s := []byte{}
+	for i = len(salt); i > 0; i -= MIXCHARS {
+		if i > MIXCHARS {
+			s = append(s, sumDS...)
+		} else {
+			s = append(s, sumDS[0:i]...)
+		}
+	}
+
+	// 21
+	tmpC := sha256.New()
+	var sumC []byte
+	for i = 0; i < iterations; i++ {
+		tmpC.Reset()
+
+		if i&1 != 0 {
+			tmpC.Write(p)
+		} else {
+			tmpC.Write(sumA)
+		}
+		if i%3 != 0 {
+			tmpC.Write(s)
+		}
+		if i%7 != 0 {
+			tmpC.Write(p)
+		}
+		if i&1 != 0 {
+			tmpC.Write(sumA)
+		} else {
+			tmpC.Write(p)
+		}
+		sumC = tmpC.Sum(nil)
+		copy(sumA, tmpC.Sum(nil))
+	}
+
+	// 22
+	buf := bytes.Buffer{}
+	buf.Grow(100) // FIXME
+	buf.Write([]byte{'$', 'A', '$'})
+	rounds := fmt.Sprintf("%03d", iterations/ITERATION_MULTIPLIER)
+	buf.Write([]byte(rounds))
+	buf.Write([]byte{'$'})
+	buf.Write(salt)
+
+	buf.Write(b64_from_24bit([]byte{sumC[0], sumC[10], sumC[20]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[21], sumC[1], sumC[11]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[12], sumC[22], sumC[2]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[3], sumC[13], sumC[23]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[24], sumC[4], sumC[14]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[15], sumC[25], sumC[5]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[6], sumC[16], sumC[26]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[27], sumC[7], sumC[17]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[18], sumC[28], sumC[8]}, 4))
+	buf.Write(b64_from_24bit([]byte{sumC[9], sumC[19], sumC[29]}, 4))
+	buf.Write(b64_from_24bit([]byte{0, sumC[31], sumC[30]}, 3))
+
+	return buf.String()
+}
+
+// Checks if a MySQL style caching_sha2 authentication string matches a password
+func CheckShaPassword(pwhash []byte, password string) (bool, error) {
+	pwhash_parts := bytes.Split(pwhash, []byte("$"))
+	if len(pwhash_parts) != 4 {
+		return false, errors.New("failed to decode hash parts")
+	}
+
+	hash_type := string(pwhash_parts[1])
```

Comment:
Check `pwhash_parts` size first?
If my memory is correct, TiDB does not do `Auth()` in a separate goroutine, that means a panic in `pwhash_parts[2]` make the process exit. See https://github.com/pingcap/parser/pull/1197

Reply:
Or `defer recover` here if verify the input in advance is hard? https://blog.golang.org/defer-panic-and-recover

Reply:
Isn't that done on line 184 ?